go/plugins/postgresql/retriever.go

```diff
@@ -2,16 +2,136 @@ package postgresql
 
 import (
 	"context"
+	"fmt"
+	"slices"
+	"strconv"
+	"strings"
 
 	"github.com/firebase/genkit/go/ai"
 )
 
 // RetrieverOptions options for retriever
 type RetrieverOptions struct {
+	Filter           any
+	K                int
+	DistanceStrategy DistanceStrategy
 }
 
 // Retrieve returns the result of the query
 func (ds *docStore) Retrieve(ctx context.Context, req *ai.RetrieverRequest) (*ai.RetrieverResponse, error) {
-	// TODO: implement
-	return nil, nil
+	if req.Options == nil {
+		req.Options = &RetrieverOptions{
+			Filter:           nil,
+			K:                defaultCount,
+			DistanceStrategy: defaultDistanceStrategy,
+		}
+	}
+
+	ropt, ok := req.Options.(*RetrieverOptions)
+	if !ok {
+		return nil, fmt.Errorf("postgres.Retrieve options have type %T, want %T", req.Options, &RetrieverOptions{})
+	}
+
+	ereq := &ai.EmbedRequest{
+		Documents: []*ai.Document{req.Query},
+		Options:   ds.config.EmbedderOptions,
+	}
+	eres, err := ds.config.Embedder.Embed(ctx, ereq)
+	if err != nil {
+		return nil, fmt.Errorf("postgres.Retrieve retrieve embedding failed: %v", err)
+	}
+	res, err := ds.query(ctx, ropt, eres.Embeddings[0].Embedding)
+	if err != nil {
+		return nil, fmt.Errorf("googlecloudsql.postgres.Retrieve failed to execute the query: %v", err)
+	}
+	return res, nil
+}
+
+func (ds *docStore) query(ctx context.Context, ropt *RetrieverOptions, embbeding []float32) (*ai.RetrieverResponse, error) {
+	res := &ai.RetrieverResponse{}
+
+	query := ds.buildQuery(ropt, embbeding)
+	rows, err := ds.engine.Pool.Query(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+	fieldDescriptions := rows.FieldDescriptions()
+	columnNames := make([]string, len(fieldDescriptions))
+
+	for i, fieldDescription := range fieldDescriptions {
+		columnNames[i] = fieldDescription.Name
+	}
+
+	for rows.Next() {
+		values := make([]interface{}, len(columnNames))
+		valuesPrt := make([]interface{}, len(columnNames))
+
+		for i := range columnNames {
+			valuesPrt[i] = &values[i]
+		}
+		if err := rows.Scan(valuesPrt...); err != nil {
+			return nil, fmt.Errorf("scan row failed: %v", err)
+		}
+
+		meta := make(map[string]any, ropt.K)
+		var content []*ai.Part
+		for i, col := range columnNames {
+			if (len(ds.config.MetadataColumns) > 0 && !slices.Contains(ds.config.MetadataColumns, col)) &&
+				ds.config.ContentColumn != col &&
+				ds.config.MetadataJSONColumn != col {
+				continue
+			}
+
+			if ds.config.ContentColumn == col {
+				content = append(content, ai.NewTextPart(values[i].(string)))
+			}
+
+			if ds.config.MetadataJSONColumn == col {
```

Comment:
MetadataJSON should be the base of the metadata dict. See https://github.com/averikitsch/langchaingo/blob/ac70ea0e7d0b8ee8b9ba2c9bf2631aa8bbb329f6/vectorstores/cloudsql/vectorstore.go#L218